core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,445 @@
+}
+
+type BundlePool struct {
+	config    Config
+	mevConfig miner.MevConfig
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	simulator BundleSimulator
+
+	sequencerClients map[string]*rpc.Client
+	deliverBundleCh  chan *types.SendBundleArgs
+	exitCh           chan struct{}
+}
+
+func New(config Config, mevConfig miner.MevConfig) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:           config,
+		mevConfig:        mevConfig,
+		bundles:          make(map[common.Hash]*types.Bundle),
+		bundleHeap:       make(BundleHeap, 0),
+		sequencerClients: make(map[string]*rpc.Client),
+		deliverBundleCh:  make(chan *types.SendBundleArgs),
+		exitCh:           make(chan struct{}),
+	}
+	if !pool.mevConfig.SentryEnabled {
+		return pool
+	}
+	for _, v := range mevConfig.Sequencers {
+		pool.register(v)
+	}
+	if len(pool.sequencerClients) == 0 {
+		log.Error("No valid sequencers")
+	}
+	go pool.deliverLoop()
+
+	return pool
+}
+
+func (p *BundlePool) deliverLoop() {
+	for {
+		select {
+		case bundle := <-p.deliverBundleCh:
+			for url, cli := range p.sequencerClients {
+				cli := cli
+				url := url
+				go func() {
+					var hash common.Hash
+					err := cli.CallContext(context.Background(), &hash, "eth_sendBundle", *bundle)
+					if err != nil {
+						log.Error("failed to deliver bundle to sequencer", "url", url, "err", err)
+					}
+				}()
+			}
+		case <-p.exitCh:
+			log.Warn("the deliverBundleCh is closed")
+			return
+		}
+	}
+}
+
+func (p *BundlePool) register(url string) error {
+	var sequencerClient *rpc.Client
+	if url != "" {
+		var err error
+		sequencerClient, err = rpc.Dial(url)
+		if err != nil {
+			log.Error("failed to dial sequencer", "url", url, "err", err)
+			return err
+		}
+		p.sequencerClients[url] = sequencerClient
+	}
+	return nil
+}
+
+func (p *BundlePool) SetBundleSimulator(simulator BundleSimulator) {
+	p.simulator = simulator
+}
+
+func (p *BundlePool) Init(gasTip *big.Int, head *types.Header, reserve txpool.AddressReserver) error {
+	return nil
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle, originBundle *types.SendBundleArgs) error {
+	if p.simulator == nil {
+		return ErrSimulatorMissing
+	}
+
+	if bundle.MinTimestamp > uint64(time.Now().Unix()+maxMinTimestampFromNow) {
+		return ErrBundleTimestampTooHigh
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+
+	if p.mevConfig.SentryEnabled {
+		p.deliverBundleCh <- originBundle
+	}
+
+	if price.Cmp(p.minimalBundleGasPrice()) < 0 && p.slots+numSlots(bundle) > p.config.GlobalSlots {
```

Comment:
This filter condition is after line 188, which causes some transactions that do not meet the price conditions and exceed the slots to be propagated to the sequencer. Is this expected behavior?

Reply:
No, it is unreasonable logic. We should deliver bundle to receiver after these conditions. Fixed, please review again.